feat: proposal for version 1.2

daKmoR opened the pull request:
- Make it something for people to start with
- Explain that it is versioned
- Show example for version 1.2
- Show changelog for version 1.2
- Explain changes
- Add Version History
- Add existing tools

justinfagnani commented on README.md:
```diff
@@ -1,11 +1,18 @@
+Its primary goal is to offer an overview of the full public API of multiple custom elements.
 
-This effort is in the initial discussion and requirements-gathering phase. Please join the discussion in [Issues](https://github.com/webcomponents/custom-elements-json/issues).
+Key points for web components are:
```
"Key points" -> how about "Key API surfaces"?

daKmoR replied:
thx adjusted 👍

justinfagnani commented on README.md:
```diff
@@ -1,11 +1,18 @@
+- events
+- methods
+- [slots](https://html.spec.whatwg.org/multipage/scripting.html#the-slot-element) 
+- [CSS custom properties](https://drafts.csswg.org/css-variables/#defining-variables)
```
Is "CSS custom variables" the more correct name?

daKmoR replied:
> A custom property is any property whose name starts with two dashes

I think the term is correct... I always say css variables and then correct myself to css custom properties as I think that is the spec name... but I could be totally wrong 🙈

bengfarrell replied:
I'm being super picky, but you asked :) "Its primary goal is to offer...."
"Its" feels awkward here, given how short the doc is you're likely talking about custom-elements.json, but you could also mean this document maybe. I'd be more direct and say "The primary goal of custom-elements.json is to offer..."

justinfagnani commented on README.md:
```diff
@@ -1,11 +1,18 @@
-# custom-elements-json
-A file format for describing custom elements
+# custom-elements.json
+A file format for describing custom elements.
```
That's not a sentence, so it doesn't need a period :)

justinfagnani commented on README.md:
```diff
@@ -16,16 +23,178 @@ There have been, and are currently, several efforts in this area:
+
+### Version 2
+
+The main goal if this version is to address all key points of web-components.
```
"address all key points" -> "document key API surfaces"?

justinfagnani commented on README.md:
```diff
@@ -16,16 +23,178 @@ There have been, and are currently, several efforts in this area:
+It is fully based on [Version 1](#version-1). 
+It adds all mentioned key points for web components.
+It on purpose does not define a type system. This is left to the next iteration.
+Therefore "type" should be seen as a type hint which can be presented to the end-user as documentation.
```
I think we should put in some work to define the type field better

daKmoR replied:
added them via "Reference import" now 👍 
please take a look 🤗

justinfagnani commented on README.md:
````diff
@@ -16,16 +23,178 @@ There have been, and are currently, several efforts in this area:
+
+Currently implemented in VS Code. See the details for [vscode-html-languageservice](https://github.com/microsoft/vscode-html-languageservice/blob/master/docs/customData.md#custom-data-for-html-language-service).
+
+```json
````
Not sure we need the inline demo for version one if it's going to not be considered the current version when this is published.

daKmoR replied:
Version history can be quite informative therefore I think it's worth the effort but it doesn't need to be in the main readme for sure.

daKmoR replied:
moved to a separate file 👍

justinfagnani commented on README.md:
````diff
@@ -16,16 +23,178 @@ There have been, and are currently, several efforts in this area:
+
+Following is an example which shows all available capabilities.
+
+```json
````
I think rather than an inline demo, we should build out the JSON Schema. I'd personally prefer to specify it in TypeScript, then use: https://github.com/YousefED/typescript-json-schema

daKmoR replied:
In my experience having a demo json is the best way of "teaching" it to others. Having a json schema seems to even confuse people as often an assumption is made that their `custom-elements.json` has to be written in the same way. This may scare people off as a schema file looks kinda complicated.

And I can totally relate as for my learning type having a demo is the best way - for all other things I have to construct the demo in my mind first which usually comes with quite a cognitive load.

imho having a schema is an implementation detail and can be linked but should NOT be the main part of this page. Let's make the main readme more an explainer and have implementation details for a spec in a separate file.

runem replied:
My suggestion here might be too `web-component-analyzer`-specific, but I'm currently working on an online playground that analyzes custom elements and shows the corresponding `custom-elements.json` output: https://runem.github.io/web-component-analyzer/?format=json (inspired by https://polymer.github.io/analyzer-playground/)

I think the playground provides a really good way to understand the custom-elements.json format, and therefore I think it would be a nice addition to this discussion.

daKmoR replied:
@runem that is suupper awesome 🤗 I was really hoping we can get something like this at some point... but it's already there 💪

daKmoR replied:
added a link to the playground 👍

bengfarrell replied:
"The main goal if this version is to document key API surfaces of web components." 
Change "if" to "of"

justinfagnani commented:
Thanks for getting this started!

I think we should probably start work on a branch, so it doesn't appear "published" to the repo quite yet, and also start with JSON schema so we have exact details on what the files mean.

cdata commented on README.md:
```diff
@@ -16,16 +23,178 @@ There have been, and are currently, several efforts in this area:
+          "type": "color"
+        }
+      ],
+      "cssShadowParts": [
```
Since "CSS Custom Properties" was shortened (presumably for brevity), would this be better as "CSS Parts"?

daKmoR replied:
I would rather go for renaming `cssProperties` to `cssCustomProperties`? better be explicit? what do you think?

cdata replied:
Yah, I would go one way or the other. Concise or pedantic, not a mix of both :)

daKmoR replied:
thx adjusted to `cssCustomProperties` 👍

daKmoR commented:
update time
- renamed it to version 1.2 (as vscode already has version 1.1) 
- added a `schema.ts` and `schema.json`
- added path for tags (to get to the source file)
- renamed `cssProperties` to `cssCustomProperties`
- added privacy for properties and methods
- added demos with name, description, url
- added types per "Reference"

for example

```json
      "properties": [
        {
          "name": "age",
          "type": {
            "name": "PersonAge",
            "import": "demo-wc-card/index.js"
          }
        }
      ],
```

so the type for `age` can be translated to `import { PersonAge } from 'demo-wc-card/index.js`.

in ts it is probably clearer
```ts
export interface PropertyDoc extends FieldDoc {
  // [...]
  /**
   * The type of the property
   *
   * If the type is built-in, then it is a string, e.g. string|number|...,
   * If the type is define in a module then reference to it.
   */
  type?: Reference | string;
}

/**
 * A reference to an export of a module.
 *
 * All references are required to be publicly accessible, so the canonical
 * representation of a reference it the export it's available from.
 */
export interface Reference {
  /**
   * Name of the type
   */
  name: string;

  /**
   * Bare module path to the file offering the named export
   * Can be translated to import { name } from import
   * @example
   * import { foo } from 'my-package/index.js';
   */
  import?: string;
}
```

bengfarrell commented on README.md:
```diff
@@ -1,31 +1,201 @@
-## Linting
+### Documentation Playground
+
+With full information about the public api of a web component tools can provide UIs to trigger/execute these apis.
```
"public api of a web component tools can" - I think you need a comma to break this up: "public api of a web component, tools can..."

bengfarrell commented on README.md:
```diff
@@ -1,31 +1,201 @@
+#### Playground
 
-Linters should be able to produce warnings based on custom element defintions, such as warning if unknown elements are used in HTML templates.
+If you are looking for a way to find out how your code will be represented in `custom-elements.json` you can visit the [playground of web-component-analyzer](https://runem.github.io/web-component-analyzer/?format=json). Paste you web component code in and see which data currently can get extracted automatically.
```
"Paste your web component code" not "Paste you web component code"

georges-gomes commented on schema.ts:
```diff
@@ -0,0 +1,222 @@
+  /**
+   * Relative path to the js file which registers this web component
+   */
+  path?: string;
```
Should we call it 'module' ? like in 'pkg.module'?
Just to avoid people shooting UMDs here...

steel commented:
Not sure if this is the correct place to contribute but I've been getting `custom-elements.json` deployed with Storybook and wasn't happy with the current state of JSDoc generators for vanilla custom elements so I wrote my own. https://github.com/steel/custom-element-analyzer

It specifically uses JSDoc under the hood for all the parsing for maximum compatibilty. Hope it helps anyone one else who's using vanilla custom elements. Though it might work in other contexts as long as JSDoc can parse it.

web-padawan commented:
@steel out of curiosity, why did you choose to write your own tool instead of using this one?
https://github.com/runem/web-component-analyzer

steel commented:
> @steel out of curiosity, why did you choose to write your own tool instead of using this one?
> https://github.com/runem/web-component-analyzer

That one implements its own tree walker rather than using JSDoc to do the parsing for JSDoc blocks. I've been playing around with it for a few weeks and it's missing many JSDoc native features. I've also tried to submit patches without success. I'm using JSDoc in vanilla js custom elements exclusively so I wrote my own to tap into the JSDoc plugin infrastructure to handle the custom tagging.

runarberg commented on README.md:
```diff
@@ -1,31 +1,201 @@
+- Added tag slots with name, description
+  - Lists all [slots](https://html.spec.whatwg.org/multipage/scripting.html#the-slot-element) in the shadow dom like `<slot name="header"></slot>`
+  - The default slot is represented via an empty name `<slot></slot>`
+- Added tag cssProperties with name, description, type
```
This field is called `cssCustomProperties` in the schema and in the example above

daKmoR commented:
superseded by the new format on master 👍
